File: 20-valid-parentheses/valid-parentheses.cpp
class Solution {
public:
    bool isValid(string s) {
        stack<char> st;
        for (int i = 0; i < s.size(); i++) {
            char ch = s[i];
            if (ch == '[' || ch == '{' || ch == '(') {
                st.push(ch);
            } else { // closing bracket
                if (!st.empty() && ((ch == ')' && st.top() == '(') ||
                                    (ch == ']' && st.top() == '[') ||
                                    (ch == '}' && st.top() == '{'))) {
                    st.pop();
                } else {
                    return false;
                }
            }
        }
        return st.empty();
    }
};

File: 20-valid-parentheses/valid-parentheses.java
import java.util.HashMap;
import java.util.Stack;

class Solution {

    public boolean isValid(String s) {

        HashMap<Character, Character> map = new HashMap<>();

        map.put(')', '(');
        map.put('}', '{');
        map.put(']', '[');

        Stack<Character> stack = new Stack<>();

        for (char ch : s.toCharArray()) {

            if (ch == '(' || ch == '{' || ch == '[') {
                stack.push(ch);
            } 
            else {

                if (stack.isEmpty()) {
                    return false;
                }

                char top = stack.pop();

                if (top != map.get(ch)) {
                    return false;
                }
            }
        }

        return stack.isEmpty();
    }
}